juju/client/connection.py

```diff
@@ -446,16 +456,30 @@ def connect_params(self):
         tasks = [self.loop.create_task(_try_endpoint(endpoint, cacert,
                                                      0.1 * i))
                  for i, (endpoint, cacert) in enumerate(endpoints)]
-        for task in asyncio.as_completed(tasks, loop=self.loop):
-            try:
-                result = await task
-                break
-            except ConnectionError:
-                continue  # ignore; try another endpoint
-        else:
-            raise errors.JujuConnectionError(
-                'Unable to connect to any endpoint: {}'.format(', '.join([
-                    endpoint for endpoint, cacert in endpoints])))
+        for attempt in range(self._retries + 1):
+            for task in asyncio.as_completed(tasks, loop=self.loop):
+                try:
+                    result = await task
+                    break
+                except ConnectionError:
+                    continue  # ignore; try another endpoint
+            else:
+                _endpoints_str = ', '.join([endpoint
+                                            for endpoint, cacert in endpoints])
+                if attempt < self._retries:
+                    log.debug('Retrying connection to endpoints: {}; '
+                              'attempt {} of {}'.format(_endpoints_str,
+                                                        attempt + 1,
+                                                        self._retries + 1))
```

Comment:
got a little crazy with the `x + 1` here didn't ya?  this is gonna say `attempt [1|2|3] of 4` when `retries = 3`.

Reply:
One initial attempt, plus 3 retries equals 4 total attempts.  *shrug*  I guess it depends on whether you think the `retries` param should be "total number of times to try" vs "number of times to try again if it fails".

Reply:
aaaah, ok.  i retract my objection, but know that nobody but you thinks this way.